k8srm-prototype/pkg/api/claim_types.go

```diff
@@ -355,16 +263,79 @@ type DeviceConfigReference struct {
 	Name string `json:"name"`
 }
 
-// DevicePoolAllocation contains the pool and number of selected devices.
-type DevicePoolAllocation struct {
+// DeviceAllocation contains an individual device allocation result, including
+// per-device resource allocations, when applicable.
+type DeviceAllocation struct {
+	// DevicePoolName is the name of the DevicePool to which this
+	// device belongs.
+	// +required
+	DevicePoolName string `json:"devicePoolName"`
 
-	// DevicePoolName contains the name of the DevicePool for this
-	// allocation.
+	// DeviceName contains the name of the allocated Device.
 	// +required
-	DevicePoolName string `json:"devicePoolName,omitempty"`
+	DeviceName string `json:"deviceName,omitempty"`
 
-	// DeviceCount contains the number of devices allocated from the
-	// pool to satisfy this claim.
+	// Allocations contain the resource allocations from this device,
+	// for the claim. Note that this may only satisfy part of the claim.
+	// Also, because devices may allocate some resources in blocks, this
+	// may even be larger than the requests or limits in the claim.
+	// +optional
+	Allocations []ResourceAllocation `json:"allocations,omitempty"`
+}
+
+// ResourceAllocation contains the per-device resource allocations.
+type ResourceAllocation struct {
+	// Name is the resource name/string for this allocation.
 	// +required
-	DeviceCount int `json:"deviceCount,omitempty"`
+	Name string `json:"name"`
+
+	// Amount is the amount of resource allocated for this claim.
+	// +required
+	Allocation resource.Quantity `json:"allocation,inline"`
+
+	// If we ever need to support intra-device topology (for example,
+	// if standard node compute becomes a device), then we may also
+	// include topology assignments in here.
+}
+
+type DeviceIP struct {
+	// IP is the IP address assigned to the device
+	IP string `json:"ip,omitempty"`
+}
+
+// DeviceStatus contains the status of an allocated result, if the driver
+// chooses to report it. This may include driver-specific information.
+type DeviceStatus struct {
+	// DevicePoolName is the name of the DevicePool to which this
+	// device belongs. The driver for that device pool owns this
+	// entry.
+	// +required
+	DevicePoolName string `json:"devicePoolName"`
+
+	// DeviceName contains the name of the allocated Device.
+	// +required
+	DeviceName string `json:"deviceName,omitempty"`
+
+	// Conditions contains the latest observation of the device's state.
+	Conditions []metav1.Condition `json:"conditions"`
+
+	// DeviceIP contains the IP allocated for the device, if appropriate.
+	// +optional
+	DeviceIP *string `json:"deviceIP,omitempty"`
```

Comment:
Don't need this, just keep the plural

Reply:
@thockin I understand the intent and the requirement but I'm worry if this becomes an slippery slope and later we have to add Mac Address or more network related fields, or other devices start to ask for new fields ... my question is if we need to be specific or can we use a `DeviceState map[string][]string` and then create a contention of using PodIPs to represent the assigned IPs

Reply:
That's the "DeviceInfo" below. I called out IP as "special". Another option is if we have classes of devices with common config we can create well-known structures for them. This allows logic to be built on top of the status that works across drivers. But in many cases, we just have no idea what this will look like. In the netdev case, we have some idea.

Reply:
I think this is a second-order decision, but I see both sides of it.  It's conventient to have known attributes here to avoid unnecessary indirections, but it is a slippery slope.